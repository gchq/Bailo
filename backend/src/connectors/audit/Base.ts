import { Request } from 'express'

import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../models/File.js'
import { InferenceDoc } from '../../models/Inference.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { ResponseInterface } from '../../models/Response.js'
import { ReviewInterface } from '../../models/Review.js'
import { SchemaDoc, SchemaInterface } from '../../models/Schema.js'
import { TokenDoc } from '../../models/Token.js'
import { ModelSearchResult } from '../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../types/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateModelCard: { typeId: 'CreateModelCard', description: 'Model Card Created', auditKind: AuditKind.Create },
  ViewModelCard: { typeId: 'ViewModelCard', description: 'Model Card Viewed', auditKind: AuditKind.View },
  ViewModelCardRevisions: {
    typeId: 'ViewModelCardRevisions',
    description: 'Model Card Revisions Viewed',
    auditKind: AuditKind.View,
  },
  UpdateModelCard: { typeId: 'UpdateModelCard', description: 'Model Card Updated', auditKind: AuditKind.Update },

  CreateFile: { typeId: 'CreateFile', description: 'File Information Created', auditKind: AuditKind.Create },
  ViewFile: { typeId: 'ViewFile', description: 'File Downloaded', auditKind: AuditKind.View },
  ViewFiles: { typeId: 'ViewFiles', description: 'File Information Viewed', auditKind: AuditKind.View },
  DeleteFile: { typeId: 'DeleteFile', description: 'File Information Deleted', auditKind: AuditKind.Delete },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'DeleteRelease', description: 'Release Deleted', auditKind: AuditKind.Delete },
  ViewReleases: { typeId: 'ViewReleases', description: 'Releases Viewed', auditKind: AuditKind.View },

  CreateUserToken: { typeId: 'CreateUserToken', description: 'Token Created', auditKind: AuditKind.Create },
  ViewUserTokens: { typeId: 'ViewUserToken', description: 'Token Viewed', auditKind: AuditKind.View },
  DeleteUserToken: { typeId: 'DeleteUserToken', description: 'Token Deleted', auditKind: AuditKind.Delete },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  ViewAccessRequests: {
    typeId: 'ViewAccessRequests',
    description: 'Access Requests Viewed',
    auditKind: AuditKind.View,
  },

  SearchReviews: {
    typeId: 'SearchReviews',
    description: 'Reviews Searched',
    auditKind: AuditKind.Search,
  },
  CreateReviewResponse: {
    typeId: 'CreateReviewResponse',
    description: 'Review Response Created',
    auditKind: AuditKind.Create,
  },
  UpdateReviewResponse: {
    typeId: 'UpdateReviewResponse',
    description: 'Review Response Updated',
    auditKind: AuditKind.Update,
  },

  CreateSchema: { typeId: 'CreateSchema', description: 'Schema Created', auditKind: AuditKind.Create },
  SearchSchemas: { typeId: 'SearchedSchemas', description: 'Schemas Searched', auditKind: AuditKind.Search },
  ViewSchema: { typeId: 'ViewSchema', description: 'Schema Viewed', auditKind: AuditKind.View },
  DeleteSchema: { typeId: 'DeleteSchema', description: 'Schema Deleted', auditKind: AuditKind.Delete },
  UpdateSchema: { typeId: 'UpdateSchema', description: 'Schema Updated', auditKind: AuditKind.Update },

  ViewModelImages: { typeId: 'ViewModelImages', description: 'Model Images Viewed', auditKind: AuditKind.View },

  CreateInference: { typeId: 'CreateInference', description: 'Inference Service Created', auditKind: AuditKind.Create },
  UpdateInference: { typeId: 'UpdateInference', description: 'Inference Service Updated', auditKind: AuditKind.Update },
  ViewInference: { typeId: 'ViewInference', description: 'Inference Service Viewed', auditKind: AuditKind.View },
  ViewInferences: { typeId: 'ViewInferences', description: 'Inferences Viewed', auditKind: AuditKind.View },

  CreateExport: { typeId: 'CreateExport', description: 'Model Exported', auditKind: AuditKind.Create },
  CreateImport: { typeId: 'CreateImport', description: 'Model Imported', auditKind: AuditKind.Create },

  ViewResponses: {
    typeId: 'ViewResponses',
    description: 'View a list of responses',
    auditKind: AuditKind.View,
  },
  CreateResponse: {
    typeId: 'CreateResponse',
    description: 'Review or comment responses created',
    auditKind: AuditKind.Create,
  },
  UpdateResponse: {
    typeId: 'UpdateResponse',
    description: 'Updated a comment or review response',
    auditKind: AuditKind.Update,
  },
} as const
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onUpdateModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCardRevisions(req: Request, modelId: string, modelCards: ModelCardInterface[])

  abstract onCreateFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFiles(req: Request, modelId: string, files: FileInterface[])
  abstract onDeleteFile(req: Request, modelId: string, fileId: string)

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onViewReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateCommentResponse(req: Request, response: ResponseInterface)
  abstract onCreateReviewResponse(req: Request, response: ResponseInterface)

  abstract onViewResponses(req: Request, responseInterfaces: ResponseInterface[])
  abstract onUpdateResponse(req: Request, responseId: string)

  abstract onCreateUserToken(req: Request, token: TokenDoc)
  abstract onViewUserTokens(req: Request, tokens: TokenDoc[])
  abstract onDeleteUserToken(req: Request, accessKey: string)

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onViewAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[])

  abstract onSearchSchemas(req: Request, schemas: SchemaInterface[])
  abstract onCreateSchema(req: Request, schema: SchemaInterface)
  abstract onViewSchema(req: Request, schema: SchemaInterface)
  abstract onDeleteSchema(req: Request, schemaId: string)
  abstract onUpdateSchema(req: Request, schema: SchemaDoc)

  abstract onCreateInference(req: Request, inference: InferenceDoc)
  abstract onUpdateInference(req: Request, inference: InferenceDoc)
  abstract onViewInference(req: Request, inference: InferenceDoc)
  abstract onViewInferences(req: Request, inference: InferenceDoc[])

  abstract onViewModelImages(
    req: Request,
    modelId: string,
    images: { repository: string; name: string; tags: string[] }[],
  )

  abstract onCreateS3Export(req: Request, modelId: string, semvers?: string[])
  abstract onCreateImport(
    req: Request,
    mirroredModelId: string,
    sourceModelId: string,
    modelCardVersions: number[],
    exporter: string,
  )

  abstract onError(req: Request, error: BailoError)

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
