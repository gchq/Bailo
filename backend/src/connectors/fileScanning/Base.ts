import { FileInterface } from '../../models/File.js'
import { ScanInterface } from '../../models/Scan.js'
import log from '../../services/log.js'
export type FileScanResult = Pick<
  ScanInterface,
  'toolName' | 'scannerVersion' | 'state' | 'isInfected' | 'viruses' | 'lastRunAt'
>

export const ScanState = {
  NotScanned: 'notScanned',
  InProgress: 'inProgress',
  Complete: 'complete',
  Error: 'error',
} as const
export type ScanStateKeys = (typeof ScanState)[keyof typeof ScanState]

export type FileScanningConnectorInfo = Pick<FileScanResult, 'toolName' | 'scannerVersion'>

export abstract class BaseFileScanningConnector {
  abstract toolName: string
  abstract version: string | undefined

  abstract init()
  abstract scan(file: FileInterface): Promise<FileScanResult[]>

  info(): FileScanningConnectorInfo {
    return { toolName: this.toolName, scannerVersion: this.version }
  }

  async scanError(message: string, context?: object): Promise<FileScanResult[]> {
    const scannerInfo = await this.info()
    log.error({ ...context, ...scannerInfo }, message)
    return [
      {
        ...scannerInfo,
        state: ScanState.Error,
        lastRunAt: new Date(),
      },
    ]
  }
}
