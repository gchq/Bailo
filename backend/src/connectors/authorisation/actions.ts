import { TokenActions } from '../../models/Token.js'

export const ModelAction = {
  Create: 'model:create',
  View: 'model:view',
  Update: 'model:update',
  Write: 'model:write',
} as const
export type ModelActionKeys = (typeof ModelAction)[keyof typeof ModelAction]

export const ReleaseAction = {
  Create: 'release:create',
  View: 'release:view',
  Delete: 'release:delete',
  Update: 'release:update',
} as const
export type ReleaseActionKeys = (typeof ReleaseAction)[keyof typeof ReleaseAction]

export const AccessRequestAction = {
  Create: 'access_request:create',
  View: 'access_request:view',
  Update: 'access_request:update',
  Delete: 'access_request:delete',
} as const
export type AccessRequestActionKeys = (typeof AccessRequestAction)[keyof typeof AccessRequestAction]

export const SchemaAction = {
  Create: 'schema:create',
  Delete: 'schema:delete',
  Update: 'schema:update',
} as const
export type SchemaActionKeys = (typeof SchemaAction)[keyof typeof SchemaAction]

export const FileAction = {
  Delete: 'file:delete',
  Upload: 'file:upload',
  // 'view' refers to the ability to see metadata about the file.  'download' lets the user view the file contents.
  View: 'file:view',
  Download: 'file:download',
  Update: 'file:update',
} as const
export type FileActionKeys = (typeof FileAction)[keyof typeof FileAction]

export const ImageAction = {
  Pull: 'image:pull',
  Push: 'image:push',
  List: 'image:list',
  Wildcard: 'image:wildcard',
  Delete: 'image:delete',
} as const
export type ImageActionKeys = (typeof ImageAction)[keyof typeof ImageAction]

export const ResponseAction = {
  Create: 'response:create',
  View: 'response:view',
  Update: 'response:update',
} as const
export type ResponseActionKeys = (typeof ResponseAction)[keyof typeof ResponseAction]

export const ActionLookup = {
  [ModelAction.Create]: TokenActions.ModelWrite.id,
  [ModelAction.View]: TokenActions.ModelRead.id,
  [ModelAction.Update]: TokenActions.ModelWrite.id,
  [ModelAction.Write]: TokenActions.ModelWrite.id,

  [ReleaseAction.Create]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.View]: TokenActions.ReleaseRead.id,
  [ReleaseAction.Delete]: TokenActions.ReleaseWrite.id,
  [ReleaseAction.Update]: TokenActions.ReleaseWrite.id,

  [AccessRequestAction.Create]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.View]: TokenActions.AccessRequestRead.id,
  [AccessRequestAction.Update]: TokenActions.AccessRequestWrite.id,
  [AccessRequestAction.Delete]: TokenActions.AccessRequestWrite.id,

  [SchemaAction.Create]: TokenActions.SchemaWrite.id,
  [SchemaAction.Delete]: TokenActions.SchemaWrite.id,
  [SchemaAction.Update]: TokenActions.SchemaWrite.id,

  [FileAction.Delete]: TokenActions.FileWrite.id,
  [FileAction.Upload]: TokenActions.FileWrite.id,
  [FileAction.View]: TokenActions.FileRead.id,
  [FileAction.Download]: TokenActions.FileRead.id,
  [FileAction.Update]: TokenActions.FileWrite.id,

  [ImageAction.Pull]: TokenActions.ImageRead.id,
  [ImageAction.Push]: TokenActions.ImageWrite.id,
  [ImageAction.List]: TokenActions.ImageRead.id,
  [ImageAction.Wildcard]: TokenActions.ImageWrite.id,
  [ImageAction.Delete]: TokenActions.ImageWrite.id,
} as const
export type ActionLookupKeys = (typeof ActionLookup)[keyof typeof ActionLookup]
