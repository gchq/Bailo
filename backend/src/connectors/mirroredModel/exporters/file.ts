import { Readable } from 'node:stream'

import { FileWithScanResultsInterface } from '../../../models/File.js'
import { ModelDoc } from '../../../models/Model.js'
import { UserInterface } from '../../../models/User.js'
import { downloadFile } from '../../../services/file.js'
import { MirrorLogData } from '../../../services/mirroredModel/mirroredModel.js'
import { addEntryToTarGzUpload, initialiseTarGzUpload } from '../../../services/mirroredModel/tarball.js'
import config from '../../../utils/config.js'
import { BadReq, Forbidden, InternalError } from '../../../utils/error.js'
import { FileAction } from '../../authorisation/actions.js'
import authorisation from '../../authorisation/index.js'
import { ScanState } from '../../fileScanning/Base.js'
import scanners from '../../fileScanning/index.js'
import { MirrorKind } from '../index.js'
import { BaseExporter, requiresInit } from './base.js'

export class FileExporter extends BaseExporter {
  protected readonly file: FileWithScanResultsInterface

  constructor(user: UserInterface, model: ModelDoc, file: FileWithScanResultsInterface, logData: MirrorLogData) {
    super(user, model, logData)
    this.file = file
  }

  getFile() {
    return this.file
  }

  protected async _init() {
    await super._init()

    const fileAuth = await authorisation.file(this.user, this.model!, this.file, FileAction.Download)
    if (!fileAuth.success) {
      throw Forbidden(fileAuth.info, {
        userDn: this.user.dn,
        modelId: this.model.id,
        fileId: this.file.id,
        ...this.logData,
      })
    }

    if (this.file.size > config.modelMirror.export.maxSize) {
      throw BadReq('Requested export is too large.', {
        size: this.file.size,
        maxSize: config.modelMirror.export.maxSize,
      })
    }

    if (scanners.info()) {
      if (!this.file.avScan || this.file.avScan.length === 0) {
        throw BadReq('The file is missing AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      } else if (this.file.avScan.some((scanResult) => scanResult.state !== ScanState.Complete)) {
        throw BadReq('The file has incomplete AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      } else if (this.file.avScan.some((scanResult) => scanResult.isInfected)) {
        throw BadReq('The file has failed AV scan(s).', { filename: this.file.name, fileId: this.file.id })
      }
    }
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    return [
      `${this.file.id}.tar.gz`,
      {
        schemaVersion: 1,
        exporter: this.user.dn,
        sourceModelId: this.model.id,
        mirroredModelId: this.model.settings.mirror.destinationModelId!,
        filePath: this.file.id,
        importKind: MirrorKind.File,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addEntryToTarGzUpload(
      this.tarStream!,
      {
        type: 'stream',
        filename: this.file.id,
        stream: (await downloadFile(this.user, this.file.id)).Body as Readable,
        size: this.file!.size,
      },
      this.logData,
    )
  }
}
