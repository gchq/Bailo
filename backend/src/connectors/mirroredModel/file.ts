import { PassThrough, Readable } from 'node:stream'

import { Headers } from 'tar-stream'

import { putObjectStream } from '../../clients/s3.js'
import { ScanState } from '../../connectors/fileScanning/Base.js'
import scanners from '../../connectors/fileScanning/index.js'
import FileModel, { FileWithScanResultsInterface } from '../../models/File.js'
import { ModelDoc } from '../../models/Model.js'
import { UserInterface } from '../../models/User.js'
import { createFilePath, downloadFile, getFileById, markFileAsCompleteAfterImport } from '../../services/file.js'
import log from '../../services/log.js'
import { addEntryToTarGzUpload, initialiseTarGzUpload } from '../../services/mirroredModel/tarball.js'
import config from '../../utils/config.js'
import { BadReq, Forbidden, InternalError } from '../../utils/error.js'
import { FileAction } from '../authorisation/actions.js'
import authorisation from '../authorisation/index.js'
import { BaseExporter, BaseImporter, BaseMirrorMetadata, requiresInit } from './base.js'
import { MirrorKind, MirrorKindKeys } from './index.js'

export type FileMirrorMetadata = BaseMirrorMetadata & { importKind: MirrorKindKeys<'File'>; filePath: string }
export type FileMirrorInformation = {
  metadata: FileMirrorMetadata
  sourcePath: string
  newPath: string
}

export class FileExporter extends BaseExporter {
  protected readonly fileId: string
  protected file: FileWithScanResultsInterface | undefined

  constructor(
    user: UserInterface,
    model: string | ModelDoc,
    file: string | FileWithScanResultsInterface,
    logData?: Record<string, unknown>,
  ) {
    super(user, model, logData)

    if (typeof file === 'string') {
      this.fileId = file
    } else {
      this.file = file
      this.fileId = this.file.id
    }
  }

  getFile() {
    return this.file
  }

  async _init() {
    await super._init()

    if (!this.file) {
      this.file = await getFileById(this.user, this.fileId, this.model)
    }
    const fileAuth = await authorisation.file(this.user, this.model!, this.file, FileAction.Download)
    if (!fileAuth.success) {
      throw Forbidden(fileAuth.info, {
        userDn: this.user.dn,
        modelId: this.modelId,
        fileId: this.fileId,
        ...this.logData,
      })
    }

    if (this.file.size > config.modelMirror.export.maxSize) {
      throw BadReq('Requested export is too large.', {
        size: this.file.size,
        maxSize: config.modelMirror.export.maxSize,
      })
    }

    if (scanners.info()) {
      if (!this.file.avScan || this.file.avScan.length === 0) {
        throw BadReq('The file is missing AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.state !== ScanState.Complete)) {
        throw BadReq('The file has incomplete AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      } else if (this.file.avScan.some((scanResult) => scanResult.isInfected)) {
        throw BadReq('The file has failed AV scan(s).', { filename: this.file.name, fileId: this.fileId })
      }
    }
  }

  protected getInitialiseTarGzUploadParams(): Parameters<typeof initialiseTarGzUpload> {
    if (!this.model) {
      throw InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        ...this.logData,
      })
    }
    return [
      `${this.fileId}.tar.gz`,
      {
        exporter: this.user.dn,
        sourceModelId: this.modelId,
        mirroredModelId: this.model.settings.mirror.destinationModelId!,
        filePath: this.fileId,
        importKind: MirrorKind.File,
      },
      this.logData,
    ]
  }

  @requiresInit
  async addData() {
    // Non-null assertion operator used due to `requiresInit` performing assertion
    await addEntryToTarGzUpload(
      this.tarStream!,
      {
        type: 'stream',
        filename: this.fileId,
        stream: (await downloadFile(this.user, this.fileId)).Body as Readable,
        size: this.file!.size,
      },
      this.logData,
    )
  }
}

export class FileImporter extends BaseImporter {
  declare protected metadata: FileMirrorMetadata

  protected bucket: string
  protected updatedPath: string
  protected extractedFile: boolean = false

  constructor(metadata: FileMirrorMetadata, logData?: Record<string, unknown>) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.File) {
      throw InternalError('Cannot parse compressed File: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }
    this.bucket = config.s3.buckets.uploads
    this.updatedPath = createFilePath(this.metadata.mirroredModelId, this.metadata.filePath)
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      if (this.extractedFile) {
        throw InternalError('Cannot parse compressed file: multiple files found.', {
          mirroredModelId: this.metadata.mirroredModelId,
          fileId: this.metadata.filePath,
          entry,
          ...this.logData,
        })
      } else {
        const foundFile = await FileModel.findOne({ path: this.updatedPath, complete: true })
        if (foundFile) {
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, ...this.logData },
            'Skipping imported file as it has already been uploaded to S3.',
          )
          // auto-drain the stream
          stream.resume()
        } else {
          await putObjectStream(this.updatedPath, stream, this.bucket)
          await markFileAsCompleteAfterImport(this.updatedPath)
          log.debug(
            { bucket: this.bucket, path: this.updatedPath, name: entry.name, ...this.logData },
            'Imported file successfully uploaded to S3.',
          )
        }
        this.extractedFile = true
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.debug({ name: entry.name, type: entry.type, ...this.logData }, 'Skipping non-file entry.')
    }
  }

  // Type resolve
  finishListener(resolve: (reason?: FileMirrorInformation) => void, _reject: (reason?: unknown) => void) {
    super.finishListener(resolve, _reject)
  }
}
