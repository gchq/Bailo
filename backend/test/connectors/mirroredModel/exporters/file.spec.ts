import { Readable } from 'node:stream'

import { beforeEach, describe, expect, test, vi } from 'vitest'

import { FileAction } from '../../../../src/connectors/authorisation/actions.js'
import { ScanState } from '../../../../src/connectors/fileScanning/Base.js'
import { FileExporter } from '../../../../src/connectors/mirroredModel/exporters/file.js'
import { MirrorKind } from '../../../../src/connectors/mirroredModel/index.js'
import { BadReq, Forbidden, InternalError } from '../../../../src/utils/error.js'

const tarballMocks = vi.hoisted(() => ({
  initialiseTarGzUpload: vi.fn(),
  addEntryToTarGzUpload: vi.fn(),
}))
vi.mock('../../../../src/services/mirroredModel/tarball.js', () => tarballMocks)

const fileServiceMocks = vi.hoisted(() => ({
  downloadFile: vi.fn(),
}))
vi.mock('../../../../src/services/file.js', () => fileServiceMocks)

const authMocks = vi.hoisted(() => ({
  default: {
    model: vi.fn(),
    file: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/authorisation/index.js', () => authMocks)

const scannersMocks = vi.hoisted(() => ({
  default: {
    info: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/fileScanning/index.js', () => scannersMocks)

const configMocks = vi.hoisted(() => ({
  default: {
    modelMirror: {
      export: {
        maxSize: 1000,
      },
    },
    log: {
      level: 'debug',
    },
    instrumentation: {
      enabled: false,
    },
  },
}))
vi.mock('../../../../src/utils/config.js', () => configMocks)

const mockUser = { dn: 'userDN' } as any
const mockModel = {
  id: 'modelId',
  settings: { mirror: { destinationModelId: 'destModelId' } },
  card: { schemaId: 'schemaId' },
} as any
const mockFile = {
  id: 'fileId',
  name: 'test.txt',
  size: 500,
  avScan: [{ state: ScanState.Complete, isInfected: false }],
} as any
const mockLogData = { extra: 'info', exporterType: 'FileExporter' }

describe('connectors > mirroredModel > exporters > FileExporter', () => {
  beforeEach(() => {
    tarballMocks.initialiseTarGzUpload.mockResolvedValue({
      tarStream: {} as any,
      gzipStream: {} as any,
      uploadStream: {} as any,
      uploadPromise: Promise.resolve(),
    })
    tarballMocks.addEntryToTarGzUpload.mockResolvedValue(undefined)
    fileServiceMocks.downloadFile.mockResolvedValue({ Body: new Readable() })
    authMocks.default.model.mockResolvedValue({ success: true })
    authMocks.default.file.mockResolvedValue({ success: true })
    scannersMocks.default.info.mockReturnValue(false)
  })

  test('constructor > sets file and logData', async () => {
    const exporter = new FileExporter(mockUser, mockModel, mockFile, mockLogData)

    expect(exporter.getFile()).toEqual(mockFile)

    expect(exporter).toMatchSnapshot()
  })

  test('_init > success with valid inputs', async () => {
    const exporter = new FileExporter(mockUser, mockModel, mockFile)

    // @ts-expect-error protected method
    await exporter._init()

    expect(authMocks.default.model).toHaveBeenCalled()
    expect(authMocks.default.file).toHaveBeenCalledWith(mockUser, mockModel, mockFile, FileAction.Download)
  })

  test('_init > throws Forbidden if file authorisation fails', async () => {
    authMocks.default.file.mockResolvedValueOnce({ success: false, info: 'no file access' })
    const exporter = new FileExporter(mockUser, mockModel, mockFile)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      Forbidden('no file access', {
        userDn: mockUser.dn,
        modelId: mockModel.id,
        fileId: mockFile.id,
        exporterType: 'FileExporter',
      }),
    )
  })

  test('_init > throws BadReq if file too large', async () => {
    const largeFile = { ...mockFile, size: 2000 }
    const exporter = new FileExporter(mockUser, mockModel, largeFile)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      BadReq('Requested export is too large.', {
        size: 2000,
        maxSize: configMocks.default.modelMirror.export.maxSize,
      }),
    )
  })

  test('_init > throws BadReq if AV scans missing', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const badFile = { ...mockFile, avScan: [] }
    const exporter = new FileExporter(mockUser, mockModel, badFile)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      BadReq('The file is missing AV scan(s).', { filename: badFile.name, fileId: badFile.id }),
    )
  })

  test('_init > throws BadReq if AV scans incomplete', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const badFile = { ...mockFile, avScan: [{ state: ScanState.InProgress, isInfected: false }] }
    const exporter = new FileExporter(mockUser, mockModel, badFile)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      BadReq('The file has incomplete AV scan(s).', { filename: badFile.name, fileId: badFile.id }),
    )
  })

  test('_init > throws BadReq if AV scans infected', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const badFile = { ...mockFile, avScan: [{ state: ScanState.Complete, isInfected: true }] }
    const exporter = new FileExporter(mockUser, mockModel, badFile)

    // @ts-expect-error protected method
    await expect(exporter._init()).rejects.toEqual(
      BadReq('The file has failed AV scan(s).', { filename: badFile.name, fileId: badFile.id }),
    )
  })

  test('getInitialiseTarGzUploadParams > returns correct params', async () => {
    const exporter = new FileExporter(mockUser, mockModel, mockFile, mockLogData)

    // @ts-expect-error protected method
    const params = exporter.getInitialiseTarGzUploadParams()

    expect(params[0]).toBe(`${mockFile.id}.tar.gz`)
    expect(params[1]).toMatchObject({
      exporter: mockUser.dn,
      sourceModelId: mockModel.id,
      mirroredModelId: mockModel.settings.mirror.destinationModelId,
      filePath: mockFile.id,
      importKind: MirrorKind.File,
    })
    expect(params[2]).toEqual(mockLogData)
  })

  test('getInitialiseTarGzUploadParams > throws if model missing', () => {
    const exporter = new FileExporter(mockUser, undefined as any, mockFile)

    // @ts-expect-error protected method
    expect(() => exporter.getInitialiseTarGzUploadParams()).toThrowError(
      InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        exporterType: 'FileExporter',
      }),
    )
  })

  test('addData > throws if not initialised', async () => {
    const exporter = new FileExporter(mockUser, mockModel, mockFile)

    expect(() => exporter.addData()).toThrowError(
      InternalError('Method `addData` called before `init()`.', { exporterType: 'FileExporter' }),
    )
  })

  test('addData > success calls addEntryToTarGzUpload', async () => {
    const exporter = new FileExporter(mockUser, mockModel, mockFile, mockLogData)
    exporter['initialised'] = true
    exporter['tarStream'] = {} as any
    exporter['gzipStream'] = {} as any
    exporter['uploadStream'] = {} as any
    exporter['uploadPromise'] = Promise.resolve()

    await exporter.addData()

    expect(fileServiceMocks.downloadFile).toHaveBeenCalledWith(mockUser, mockFile.id)
    expect(tarballMocks.addEntryToTarGzUpload).toHaveBeenCalledWith(
      exporter['tarStream'],
      expect.objectContaining({
        type: 'stream',
        filename: mockFile.id,
        size: mockFile.size,
        stream: expect.any(Readable),
      }),
      mockLogData,
    )
  })
})
