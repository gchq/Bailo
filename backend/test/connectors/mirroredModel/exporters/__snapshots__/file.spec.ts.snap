// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connectors > mirroredModel > exporters > FileExporter > constructor > sets file and logData 1`] = `
FileExporter {
  "_init": [Function],
  "addData": [Function],
  "file": {
    "avScan": [
      {
        "isInfected": false,
        "state": "complete",
      },
    ],
    "id": "fileId",
    "name": "test.txt",
    "size": 500,
  },
  "getFile": [Function],
  "getInitialiseTarGzUploadParams": [Function],
  "gzipStream": undefined,
  "initialised": false,
  "logData": {
    "exportId": "exportId",
    "exporterType": "FileExporter",
    "extra": "info",
  },
  "model": {
    "card": {
      "schemaId": "schemaId",
    },
    "id": "modelId",
    "settings": {
      "mirror": {
        "destinationModelId": "destModelId",
      },
    },
  },
  "tarStream": undefined,
  "uploadPromise": undefined,
  "uploadStream": undefined,
  "user": {
    "dn": "userDN",
  },
}
`;
